orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.cc

```diff
@@ -0,0 +1,432 @@
+                                             node_index_to_its_order_in_topological_sort_map,
+                                             candidate_output_args_map,
+                                             logger, true,
+                                             can_compromise_stashed_activation);
+      if (recompute_plan != nullptr) {
+        memory_opt_planner.AddNodeOptimizationPlan(p_node, recompute_plan);
+      }
+    }
+  }
+
+  return Status::OK();
+}
+
+void GetMemoryRecordsGroupedByNodeClusterId(const MemoryOptimizationPlanner& memory_opt_planner,
+                                            std::vector<std::pair<std::string, MemoryRecord>>& generated_records,
+                                            const NodeToClusterApplyContextMap& node_to_apply_contexts_map) {
+  // Group by node cluster id, generate memory record.
+  InlinedHashMap<std::string, MemoryRecord> records;
+  const auto& node_to_optimization_plan_map = memory_opt_planner.GetNodeToOptimizationPlanMap();
+  for (const auto& node_to_optimization_plan : node_to_optimization_plan_map) {
+    const auto& node = node_to_optimization_plan.first;
+    const auto& node_plans = node_to_optimization_plan.second;
+    const std::string node_cluster_id = memory_opt_planner.GenerateNodeClusterId(node);
+
+    bool already_exist = records.find(node_cluster_id) != records.end();
+
+    auto& record = records[node_cluster_id];
+    record.freq++;
+
+    // Collect more information for display.
+    for (auto& plan : node_plans) {
+      // Same cluster id, plans might still have different reuse_buffer pattern, so we need to collect all of them.
+      if (plan->reuse_buffers.size() > 0) {
+        const InlinedVector<size_t> output_indices = plan->GetActivationOutputIndices();
+        for (auto output_index : output_indices) {
+          if (plan->GetOptimizationType() == OptimizationType::RecomputeWithCompromise) {
+            if (plan->reuse_buffers.find(output_index) != plan->reuse_buffers.end()) {
+              record.output_port_reuse_recompute_with_compromise_count[output_index] += 1;
+            }
+          } else if (plan->GetOptimizationType() == OptimizationType::Recompute) {
+            if (plan->reuse_buffers.find(output_index) != plan->reuse_buffers.end()) {
+              record.output_port_reuse_recompute_count[output_index] += 1;
+            }
+          }
+        }
+      }
+
+      // For other infos that are guaranteed identity by cluster id, just skip collecting.
+      if (already_exist) {
+        continue;
+      }
+
+      if (plan->GetOptimizationType() == OptimizationType::RecomputeWithCompromise) {
+        record.recompute_with_compromise_subgraph_str =
+            dynamic_cast<NodeRecomputePlan*>(plan.get())->GetNodesInTopoOrderStr();
+      } else if (plan->GetOptimizationType() == OptimizationType::Recompute) {
+        record.recompute_subgraph_str = dynamic_cast<NodeRecomputePlan*>(plan.get())->GetNodesInTopoOrderStr();
+      }
```

Comment:
What is needed here is a virtual function.
dybnamic_cast is expensive and works only on classes with virtual tables, so make that function virtual, so you do not need cast.

Reply:
This might still be an abstraction issue I have not solved well, while at this moment, `GetNodesInTopoOrderStr` is not a common interface defined in the base NodeOptimizationPlan class, so here introduced a dynamic cast, let me think about if there is possibly a better way doing this